libs/platform/ebpf_epoch.c

```diff
@@ -589,140 +566,219 @@ _ebpf_epoch_release_free_list(_Inout_ ebpf_epoch_cpu_entry_t* cpu_entry, int64_t
+        MemoryBarrier();
+        _ebpf_epoch_compute_release_epoch_message.message.compute_epoch.epoch = _ebpf_current_epoch;
+        KeInitializeEvent(&_ebpf_epoch_compute_release_epoch_message.completion_event, NotificationEvent, false);
+        KeSetTargetProcessorDpc(&_ebpf_epoch_compute_release_epoch_dpc, 0);
+        KeInsertQueueDpc(&_ebpf_epoch_compute_release_epoch_dpc, &_ebpf_epoch_compute_release_epoch_message, NULL);
+    } else {
+        _ebpf_epoch_skipped_timers++;
+        LARGE_INTEGER due_time;
+        due_time.QuadPart = -(EBPF_EPOCH_FLUSH_DELAY_IN_NANOSECONDS / EBPF_NANO_SECONDS_PER_FILETIME_TICK);
+        KeSetTimer(&_ebpf_epoch_compute_release_epoch_timer, due_time, &_ebpf_epoch_timer_dpc);
+    }
 }
 
-static _Requires_lock_held_(cpu_entry->lock) void _ebpf_epoch_arm_timer_if_needed(ebpf_epoch_cpu_entry_t* cpu_entry)
+/**
+ * @brief Worker for the ebpf epoch inter-CPU messaging system.
+ * It handles the following messages:
+ * 1. COMPUTE_EPOCH: Compute the minimum epoch of all threads on the CPU and send the message to the next CPU.
+ * If this is the last CPU, then set the _ebpf_release_epoch to the minimum epoch and
+ * send a message to the first CPU to rearm the timer.
+ * 2. REARM_TIMER: Remove all entries from the per-CPU free list that have an epoch that is before released_epoch,
+ * rearm the timer if needed, and send the message to the next CPU. If this is the last CPU, then stop.
+ * 3. COMPUTE_EPOCH_COMPLETE: Mark the release epoch computation as complete.
+ * 4. EXIT_EPOCH: Call ebpf_epoch_exit() on behalf of the other CPU.
+ * 5. RUNDOWN_IN_PROGRESS: Set the _ebpf_epoch_cpu_entry_t.rundown_in_progress flag to true and send the message to
+ * the next CPU. If this is the last CPU, then signal the caller that rundown is complete.
+ *
+ * If rundown is in progress, then the message is ignored.
+ *
+ * @param[in] dpc DPC that triggered this function.
+ * @param[in] context Context passed to the DPC - not used.
+ * @param[in] arg1 The ebpf_epoch_cpu_message_t to process.
+ * @param[in] arg2 Not used.
+ */
+_Function_class_(KDEFERRED_ROUTINE) _IRQL_requires_(DISPATCH_LEVEL) static void _ebpf_epoch_messenger_worker(
+    _In_ KDPC* dpc, _In_opt_ void* context, _In_opt_ void* arg1, _In_opt_ void* arg2)
 {
+    UNREFERENCED_PARAMETER(context);
+    UNREFERENCED_PARAMETER(arg2);
+    uint32_t cpu_id = ebpf_get_current_cpu();
+    uint32_t next_cpu_id = UINT32_MAX;
+    ebpf_epoch_cpu_entry_t* cpu_entry = &_ebpf_epoch_cpu_table[cpu_id];
+    ebpf_epoch_cpu_message_t* message = (ebpf_epoch_cpu_message_t*)arg1;
+
+    // If rundown is in progress, then exit immediately.
     if (cpu_entry->rundown_in_progress) {
         return;
     }
-    if (cpu_entry->timer_armed) {
-        return;
+
+    _Analysis_assume_(arg1 != NULL);
+
+    switch (message->message_type) {
+    case EBPF_EPOCH_CPU_MESSAGE_TYPE_COMPUTE_EPOCH: {
```

Comment:
This function is rather long.  Would be more readable if each case had its own helper function.
(If perf is a critical, make it an inline one)